# Obrada informacija

## Laboratorijska vježba 2

U ovoj vježbi upoznat ćete se s jednom primjenom tehnika obrade informacija u bioinformatici. Ova laboratorijska vježba nosi 4 boda. Izvješće s ove laboratorijske vježbe potrebno je predati u .pdf formatu na *Moodle*. Izvješće koje predajete se mora zvati *PrezimeIme.pdf*.

Osim biblioteka za rad s Fourierovom transformacijom (koristit ćemo samo numpy) koristit ćemo i biblioteku biopython koja sadrži puno korisnih alata iz područja bioinformatike. Mi ćemo je koristiti za jednostavnije baratanje bioinformatičkim tipovima podataka.

Biblioteka biopython dolazi s instalacijom Anaconde, ali ju je potrebno uključiti u okolinu (*environment*) koja se koristi.

Ako vježbu izvodite u Google Colab okruženju, morate instalirati biblioteku biopython. Instalaciju je potrebno izvršiti u sklopu prvog zadatka ove laboratorijske vježbe. Instalaciju izvodite sljedećim kodom:

    try:
      import google.colab
      !pip install biopython
    except ImportError:
      pass

Nakon izvođenja ovog koda, možete učitati biopython biblioteku.

1. Zadatak

Python biblioteke potrebne za laboratorijske vježbe su numpy i biopython. Uključite ih ("importirajte") i ispišite verziju svake od njih pomoću \[ime_biblioteke\].__version__.

UPUTA: Osnovna biopython biblioteka ima naziv Bio.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
try:
  import google.colab
  !pip install biopython
except ImportError:
  pass

import numpy as np
import Bio

print("Numpy verzija je " + np.__version__)
print("Biopython verzija je " + Bio.__version__)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 33.1 MB/s eta 0:00:00
Numpy verzija je 1.23.5
Biopython verzija je 1.81


2. Zadatak

Uz laboratorijske vježbe dobili ste dvije datoteke s podacima. Datoteku koja sadrži referentni genom jednog soja bakterije  Escherichia coli (escherichia_coli_reference.fasta) u FASTA formatu i datoteku koja sadrži skup očitanja dobivenih sekvenciranjem (ecoli_ILL_small.fastq) u FASTQ formatu.

Datoteke možete učitati koristeći metodu _parse()_ iz biblioteke Bio.SeqIO. Metoda _parse()_ vraća iterator koji možete pretvoriti u Python listu na sljedeći način:

    reads = list(parse("ime_datoteke", "tip_datoteke"))

Tip datoteke postavite na "fasta" ili "fastq".

Učitajte obje datoteke te ispišite broj zapisa u svakoj od njih (broj elemenata u listi). Datoteka koja sadrži referencu trebala bi imati samo jedan zapis, dok bi datoteka s očitanjima trebala sadržavati veći broj zapisa.

NAPOMENA: Ako niste sigurni kako pronaći datoteke na disku iz Jupyter notebook-a, uvijek možete provjeriti radni direktorij sljedećim naredbama:

    import os
    os.getcwd()

i promijeniti ga sa:

    os.chdir()

Ako pak radite u Google Colab okruženju, koristite upute za učitavanje datoteka s Google diska iz prve laboratorijske vježbe.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
from Bio import SeqIO

from google.colab import drive
drive.mount('/content/drive')

referenca_path = '/content/drive/My Drive/Colab Notebooks/escherichia_coli_reference.fasta'
ocitanja_path = '/content/drive/My Drive/Colab Notebooks/ecoli_ILL_small.fastq'

referenca = list(SeqIO.parse(referenca_path, "fasta"))
ocitanja = list(SeqIO.parse(ocitanja_path, "fastq"))

print("Broj zapisa u datoteci (broj elemenata u listi) koja sadrzi referencu genoma za E.coli: ", len(referenca))
print("Broj zapisa u datoteci (broj elemenata u listi) koja sadrzi ocitanja: ", len(ocitanja))

Mounted at /content/drive
Broj zapisa u datoteci (broj elemenata u listi) koja sadrzi referencu genoma za E.coli:  1
Broj zapisa u datoteci (broj elemenata u listi) koja sadrzi ocitanja:  38585


3. Zadatak

Svaki zapis koji ste učitali pomoću metode _Bio.SeqIO.parse()_ sadrži Veći broj podataka od kojih su nam bitni samo neki. Naredbom print ispišite cijeli prvi zapis iz datoteke s očitanjima i iz datoteke s referencom.

Vidjet ćete da oba zapisa (među ostalim podacima) sadrže identifikator zapisa i sekvencu. Identifikator zapisa možete dohvatiti pomoću

    zapis.id

dok sekvencu možete dohvatiti pomoću

    zapis.seq

Ispišite identifikator i sekvencu za prvo očitanje te identifikator i prvih 200 znakova za referentni genom E.coli.

NAPOMENA: Referentni genom Escherichia coli je dugačak oko 4.5 milijuna slova

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
print("Prvi zapis iz datoteke s očitanjima: ")
print(ocitanja[0])
print("\n")

print("Prvi zapis iz datoteke s referencom: ")
print(referenca)
print("\n")

print("Identifikator za prvo očitanje: " + ocitanja[0].id)
print("Sekvenca za prvo očitanje: " + str(ocitanja[0].seq))

print("\n")

print("Identifikator za referencu genoma E.coli: " + str(referenca[0].id))
print("Prvih 20 znakova refernce genoma E.coli: " + str(referenca[0].seq[:200]))

Prvi zapis iz datoteke s očitanjima: 
ID: SRR2052522.671
Name: SRR2052522.671
Description: SRR2052522.671 HWI-EAS390_0001:4:1:6915:1123/1
Number of features: 0
Per letter annotation for: phred_quality
Seq('GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAA...TGC')


Prvi zapis iz datoteke s referencom: 
[SeqRecord(seq=Seq('AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAG...TTC'), id='NC_000913.3', name='NC_000913.3', description='NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome', dbxrefs=[])]


Identifikator za prvo očitanje: SRR2052522.671
Sekvenca za prvo očitanje: GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGGGCCTCTGTATCATGCCACTCACTGCGCAATATCCGGATCAAATGC


Identifikator za referencu genoma E.coli: NC_000913.3
Prvih 20 znakova refernce genoma E.coli: AGCTTTTCATTCTGACTGCAACGGGCAATATGTCTCTGTGTGGATTAAAAAAAGAGTGTCTGATAGCAGCTTCTGAACTGGTTACCTGCCGTGAGTAAATTAAAATTTTATTGACTTAGGTCACTAAATACTTTAACCAATATAGGCATAGCGCACAGACAGATAAAAATTACAG

4. Zadatak

Da bismo sekvence DNA analizirali metodama obrade signala, moramo pojednim nukleotidima (slovima) dodijeliti brojčane vrijednosti. Napišite funkciju u Pythonu koja će primiti slovo koje predstavlja nukleotid i vratiti odgovarajuću brojčanu vrijednost. Vrijednosti dodijelite na sljedeći način:

- A =  3
- G =  2
- C = -2
- T = -3

DNA sekvence mogu sadržavati i neke druge znakove (npr. 'N' koji označava da taj nuklotid nije poznat), njima dodijelite vrijednost 0. Također se može dogoditi da nukleotidi budu označeni i malim slovima, pa vodite računa da vaša funkcija mora vratiti ispravnu vrijendost i u tom slučaju.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
def pretvorba_slova(slovo):
  if slovo.upper()=='A':
    return 3
  elif slovo.upper()=='G':
    return 2
  elif slovo.upper()=='C':
    return -2
  elif slovo.upper()=='T':
    return -3
  else:
    return 0

5. Zadatak

Upotrebite napisanu funkciju da bi od prvog očitanja i od reference kreirali signal. Izračunajte korelaciju pomoću Fourierove transformacije. Zanemarite imaginarne vrijednosti.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
import numpy as np
import scipy as sc

x1 = []
x2 = []

for slovo in ocitanja[0].seq:
  x2.append(pretvorba_slova(slovo))

for slovo in referenca[0].seq:
  x1.append(pretvorba_slova(slovo))

len1 = len(x1)
len2 = len(x2)

k_arr = range(-len2 + 1, len1)

numbers = [-3,-2,0,2,3]

avg = np.average(numbers)
std = np.std(numbers)


x1 = [(x-avg)/std for x in x1]
x2 = [(x-avg)/std for x in x2]

padding1 = [0]*(len2-1)
padding2 = [0]*(len1-1)

X1 = np.fft.fft(padding1+x1)
X2 = np.fft.fft(x2+padding2)
Cor = np.conjugate(X2)*X1
cor = np.fft.ifft(Cor)
cor = np.real(cor)

k = k_arr[cor.argmax()]
print("Korelacija po FFT:")
print(cor)
print("k={}".format(k))




Korelacija po FFT:
[-1.15384615  0.38461538 -0.19230769 ... -1.73076923 -2.30769231
 -0.76923077]
k=2324486


6. Zadatak

Ispišite duljinu reference. Koristeći metode biblioteke _numpy_, izračunajte srednju vrijednost i standardnu devijaciju duljine očitanja (uzmite u obzir sva očitanja).

Primijetit ćete da su sva očitanja jednake duljine.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
print("Duljina reference je", len(x1))

sva_ocitanja = []
for ocitanje in ocitanja:
  sva_ocitanja.append(len(ocitanje))

print("Srednja vrijednost očitanja: ", np.average(sva_ocitanja))
print("Standardna devijacija očitanja: ", np.std(sva_ocitanja))

Duljina reference je 4641652
Srednja vrijednost očitanja:  121.0
Standardna devijacija očitanja:  0.0


7. zadatak

Što ako želimo izračunati korelaciju za veći broj očitanja i istu referencu? To je tipičan slučaj u bioinformatici jer uređaji za sekvenciranje proizvode tisuće i desetke tisuća očitanja koja se potom mapiraju na istu referencu.

Ako korelaciju računamo izravno, potrebno ju je svaki put izračunati iz početka. Ako korelaciju računamo pomoću FFT-a, transformaciju za referencu potrebno je napraviti samo jednom.

Izračunajte korelacije za prvih 10 očitanja.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
for i in range(10):
  x2 = []
  for slovo in ocitanja[i].seq:
    x2.append(pretvorba_slova(slovo))
  x2 = [(x-avg)/std for x in x2]
  X2 = np.fft.fft(x2+padding2)
  Cor = np.conjugate(X2)*X1
  cor = np.fft.ifft(Cor)
  cor = np.real(cor)
  k = k_arr[cor.argmax()]
  print("Korelacija %d po FFT:" % (i+1))
  print(cor)
  print("k_{}={}".format(i+1,k))
  print("\n")



Korelacija 1 po FFT:
[-1.15384615  0.38461538 -0.19230769 ... -1.73076923 -2.30769231
 -0.76923077]
k_1=2324486


Korelacija 2 po FFT:
[ 1.73076923e+00 -3.55272330e-15 -3.65384615e+00 ... -1.73076923e+00
 -2.50000000e+00 -1.15384615e+00]
k_2=1877260


Korelacija 3 po FFT:
[-1.15384615 -2.5        -2.11538462 ... -3.46153846 -1.92307692
 -0.76923077]
k_3=557777


Korelacija 4 po FFT:
[ 1.73076923 -0.57692308 -1.15384615 ... -1.73076923 -0.96153846
 -1.15384615]
k_4=1144877


Korelacija 5 po FFT:
[ 1.73076923e+00 -1.59872122e-14 -3.07692308e+00 ...  1.92307692e-01
 -9.61538462e-01 -1.15384615e+00]
k_5=3583639


Korelacija 6 po FFT:
[ 1.15384615  1.92307692  1.73076923 ... -4.03846154 -2.5
 -1.15384615]
k_6=4051518


Korelacija 7 po FFT:
[-1.15384615e+00  9.61538462e-01  3.07692308e+00 ... -1.73076923e+00
  3.55272209e-15  7.69230769e-01]
k_7=2293706


Korelacija 8 po FFT:
[ 1.15384615  2.5        -1.34615385 ...  1.73076923  0.96153846
  1.15384615]
k_8=1011323


Korelacija 9 po FFT:
[-1

8. zadatak

Na temelju najveće vrijednosti korelacije između reference i prvog očitanja pronađite poziciju na referenci koja je najsličnija očitanju. Pozicija odgovara vrijednosti parametra k za koji je korelacija najveća.

Napišite metodu koja će primiti dva niza znakova jednake duljine, usporediti znakove na istim pozicijama i vratiti broj razlika (Hammingova udaljenost).

"Izrežite" dio reference koji je najsličniji očitanju (iste duljine kao i očitanje) i usporedite ga s očitanjem pomoću napisane funkcije.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
k=2324486
def hammingova_udaljenost(str1, str2):
    # Provjeri jesu li nizovi iste duljine
    if len(str1) != len(str2):
        raise ValueError("Nizovi moraju imati istu duljinu")

    # Inicijaliziraj brojač razlika
    udaljenost = 0

    # Iteriraj kroz znakove na istim pozicijama i usporedi ih
    for i in range(len(str1)):
        if str1[i] != str2[i]:
            print("Razlika na {}. mjestu, referenca -  {}, očitanje - {}".format((i+1),(str1[i]),(str2[i])))
            udaljenost += 1

    return udaljenost
dio_reference = referenca[0].seq[k:k+len(x2)]
print("Izrezani dio reference:", dio_reference)
print("Prvo očitanje:", str(ocitanja[0].seq))
print("Hammingova udaljenost izmedu dijela reference i prvog ocitanja:", hammingova_udaljenost(dio_reference, ocitanja[0].seq))

Izrezani dio reference: GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGCGCCGCTGCACCATGCCACTCACTGCGCAACATGCGGTGCATATGC
Prvo očitanje: GATCTGGTGACCGGGTCGCGCAAAGTGATCATCGCCATGGAACATTGCGCCAAAGATGGTTCAGCAAAAATTTTGGGCCTCTGTATCATGCCACTCACTGCGCAATATCCGGATCAAATGC
Razlika na 76. mjestu, referenca -  C, očitanje - G
Razlika na 80. mjestu, referenca -  G, očitanje - T
Razlika na 84. mjestu, referenca -  C, očitanje - T
Razlika na 86. mjestu, referenca -  C, očitanje - T
Razlika na 106. mjestu, referenca -  C, očitanje - T
Razlika na 109. mjestu, referenca -  G, očitanje - C
Razlika na 113. mjestu, referenca -  T, očitanje - A
Razlika na 114. mjestu, referenca -  G, očitanje - T
Razlika na 117. mjestu, referenca -  T, očitanje - A
Hammingova udaljenost izmedu dijela reference i prvog ocitanja: 9


9. zadatak

U datoteci "ecoli_ILL_small_aln.sam" dana su već izračunata poravnanja svih očitanja na referencu u SAM formatu. SAM je tekstualni "tab separated" format. U prvom stupcu se nalati identifikator očitanja, dok se u četvrtom stupcu nalazi pozicija na referenci na koju je očitanje najbolje poravnato (ostali stupci nas ne zanimaju). Također, datoteka s poravnanjima sadrži i nekoliko _header_ readaka kojima prvi stupac počinje sa znakom '@', njih također možete zanemariti.

Otvorite datoteku s poravnanjima i pronađite poravnanje za prvo očitanje (identifikator očitanja u datoteci s očitanjima i datoteci s poravnanjima mora biti jednak). Usporedite poziciju u datoteci sa pozicijom koju ste dobili pomoću korelacije.

UPOUTA: TSV datoteke možete otvoriti na sljedeći način:

    tsv_file = open("file_name")
    tsv_rows = csv.reader(tsv_file, delimiter="\t")

Varijabla _tsv\_rows_ će sadržavati listu redaka, a svaki redak biti lista vrijednosti (po jedna za svaki stupac).

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
import csv

tsv_file = open("/content/drive/My Drive/Colab Notebooks/ecoli_ILL_small_aln.sam")
tsv_rows = csv.reader(tsv_file, delimiter="\t")

for row in tsv_rows:
  if str(row[0])==str(ocitanja[0].id):
    print("Poravnanje za prvo očitanje u datoteci je {} za očitanje s ID: {}".format(row[3], row[0]))
    break


print("Poravnanje za prvo očitanje dobivenom korelacijom je {} za očitanje s ID: {}".format((k), ocitanja[0].id))




Poravnanje za prvo očitanje u datoteci je 2324487 za očitanje s ID: SRR2052522.671
Poravnanje za prvo očitanje dobivenom korelacijom je 2324486 za očitanje s ID: SRR2052522.671


10. zadatak

Za prvo očitanje pozicija dobivena pomoću korelacije trabala bi biti 2324486, dok je pozicija u datoteci s poravnanjima 2324487. Razlikuju se samo za 1 pa možemo zaključiti da nam je korelacija dala dobru poziciju za poravnanje.

Prisjetimo se da korelacija ne računa točno poravnanje već ju koristimo samo da bi našli kandidatne pozicije za točno računanje. Tek onda na takvim pozicijama možemo točno poravnanje izračunati pomoću algoritama dinamičkog programiranja. Ako bi primijenili dinamičko programiranje za računanje poravnanja očitanja s cijelom referencom, postupak bi bio znatno sporiji i zahtijevao bi veliku količinu radne memorije.

**Ako želite** to možete isprobati pomoću algoritama za poravnanje u biblioteci _bioparser_. Lokalno poravnanje možete izračunati metodom:

    Bio.pairwise2.align.localxx(seq1, seq2)

Za prvih 100 očitanja izračunajte korelaciju te pomoću korelacije poziciju najveće sličnosti očitanja i reference. Usporedite rezultat sa podacima u datoteci s poravnanjima. Ispišite broj očitanja za koja se dvije pozicije razlikuju za najviše 5 mjesta.

In [ ]:
# Ovo je mjesto na kojem možete izvoditi svoj kod.
sum=0
manji_od_5 = []
razlike_manje_od_5 = []
sve_razlike = []
tsv_file = open("/content/drive/My Drive/Colab Notebooks/ecoli_ILL_small_aln.sam")
tsv_rows = list(csv.reader(tsv_file, delimiter="\t"))

for i in range(100):
  x2 = []
  for slovo in ocitanja[i].seq:
    x2.append(pretvorba_slova(slovo))
  x2 = [(x-avg)/std for x in x2]
  X2 = np.fft.fft(x2+padding2)
  Cor = np.conjugate(X2)*X1
  cor = np.fft.ifft(Cor)
  cor = np.real(cor)
  k = k_arr[cor.argmax()]


  for row in tsv_rows:
    if str(row[0])==str(ocitanja[i].id):
      poravnanje_datoteka = row[3]
      razlika = k - int(poravnanje_datoteka)
      if abs(razlika) <= 5:
        #print("USAO")
        sum+=1
        manji_od_5.append(i+1)
        #print("Poravnanje za {}. očitanje u datoteci je {} za očitanje s ID: {}".format(i+1, row[3], row[0]))
        #print("Poravnanje za {}. očitanje dobivenom korelacijom je {} za očitanje s ID: {}".format(i+1, (k), ocitanja[i].id))
        #print("Razlika u poziciji poravnanja izračunatog preko korelacije i onoga iz datoteke je {} za {}. očitanje".format(abs(razlika), i+1))
        break

print("Tavkih očitanja kojima je razlika manji ili jednak 5 ima", sum)
print("To su očitanja: ", end = "")
for broj in manji_od_5:
  print(broj, end=", ")





Tavkih očitanja kojima je razlika manji ili jednak 5 ima 52
To su očitanja: 1, 2, 3, 5, 6, 8, 9, 10, 11, 12, 14, 18, 20, 21, 22, 25, 26, 27, 28, 32, 33, 35, 37, 38, 42, 43, 46, 47, 50, 52, 54, 56, 58, 59, 64, 65, 69, 70, 71, 74, 75, 79, 80, 83, 85, 86, 90, 91, 93, 95, 98, 99, 

11. ZAKLJUČAK

Očekivani broj točno pozicioniranih očitanja je 50, jer smo za sada uspješno radili samo s očitanjima na jednom lancu DNA.

Prolaskom kroz zadatke u ovoj vježbi dobili ste kratak uvod u rad s bioinformatičkim podacima i tehnikama obrade signala.